3/assets/3.6-finished-code.rs

```diff
@@ -0,0 +1,315 @@
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner == sender, "You can't set an auction for a cat you don't own");
+
+            let expiry = <system::Module<T>>::block_number() + T::BlockNumber::sa(expiry_in_minutes*12);
+            ensure!(expiry <= <system::Module<T>>::block_number() + Self::auction_duration_limit(), "The auction duration should be less than 5 days");
+
+            let new_auction = Auction {
+                kitty_id,
+                kitty_owner: owner,
+                expiry,
+                min_bid,
+                high_bid: min_bid,
+                high_bidder: sender,
+            };
+
+            <KittyAuction<T>>::insert(kitty_id, &new_auction);
+            <Auctions<T>>::insert(expiry, vec![new_auction.clone()]);
+
+            Self::deposit_event(RawEvent::Auction(kitty_id, min_bid, expiry));
+
+            Ok (())
+        }
+
+        fn bid_auction(origin, kitty_id: T::Hash, bid: T::Balance) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner != sender, "You can't bid for your own cat");
+
+            let mut auction = Self::auction_of(kitty_id).ok_or("No auction for this cat")?;
+            ensure!(<system::Module<T>>::block_number() < auction.expiry, "This auction is expired.");
+
+            ensure!(bid > auction.min_bid, "Your bid has to be greater than the minimum bid.");
+
+            ensure!(bid > auction.high_bid, "Your bid has to be greater than the highest bid.");
+
+            auction.high_bid = bid;
+            auction.high_bidder = sender.clone();
+            <KittyAuction<T>>::insert(kitty_id, &auction);
+            <Auctions<T>>::insert(auction.expiry, vec![auction.clone()]);
+
+            Self::deposit_event(RawEvent::Bid(kitty_id, auction.high_bid, auction.high_bidder));
+
+            Ok (())
+        }
+
+        fn on_finalise() {
+            let auctions = Self::auctions_expire_at(<system::Module<T>>::block_number());
+            for auction in &auctions {
+                if <system::Module<T>>::block_number() >= auction.expiry && auction.kitty_owner != auction.high_bidder {
+                    let _ = Self::_transfer_from(auction.kitty_owner.clone(), auction.high_bidder.clone(), auction.kitty_id);
+                    let _ = <balances::Module<T>>::make_transfer(&auction.high_bidder, &auction.kitty_owner, auction.high_bid);
+                }
+                <KittyAuction<T>>::remove(auction.kitty_id);
+                <Auctions<T>>::remove(<system::Module<T>>::block_number());
```

Comment:
`Remove`? Shouldn't it be `Kill`?

I have not actually tried to compile this yet, does it work for you?

Reply:
As far as I understand, if I do `<KittyAuction<T>>::kill();`, then it would clear all the other auctions for other kitties. Likewise `<Auctions<T>::kill()` would clear all the open auctions that will end at different block numbers.

Reply:
You are right, I just remember incorrectly:
https://github.com/paritytech/substrate/blob/master/srml/support/src/storage/generator.rs#L212